[zmath] Add & use expectVecApproxEqAbs and expectVecEqual.

dmurph opened the pull request:
Adding vec-based testing functions that print the failures much better. These use inner expect calls, which will throw an error on the specific mismatch instead of an opaque failure just on the line being false.

dmurph commented on libs/zmath/src/zmath.zig:
```diff
@@ -4490,18 +4490,32 @@ fn floatToIntAndBack(v: anytype) @TypeOf(v) {
+    }
+}
+
 pub fn approxEqAbs(v0: anytype, v1: anytype, eps: f32) bool {
```
note: I left this in case of dependencies & to not break people, happy to remove if desired.

hazeycode replied:
Yes let's just remove it please. We can mark the PR as a breaking change.

dmurph replied:
Done. I had to keep this for a negation check, so I renamed it to match the testing methods.

hazeycode commented:
> will throw an error on the specific mismatch instead of an opaque failure just on the line being false.

This is not what I'm seeing. The output is quite generic and I think adds no useful information. Am I missing something?

Example (without this change):
```
test
└─ run zmath-tests 79/84 passed, 1 failed, 4 skipped
error: 'zmath.test.zmath.round' failed: /Users/chris/zig/0.13.0-dev.351+64ef45eb0/files/lib/std/testing.zig:540:14: 0x102138e73 in expect (zmath-tests)
    if (!ok) return error.TestUnexpectedResult;
             ^
/Users/chris/gamedev/zig-gamedev/libs/zmath/src/zmath.zig:955:5: 0x102140f93 in test.zmath.round (zmath-tests)
    try expect(approxEqAbs(v, f32x4(-1_388_610.0, 1_388_910.0, 1_388_109.0, 2_388_610.0), 0.0));
    ^
error: while executing test 'util.test.zmath.util.mat.right', the following test command failed:
/Users/chris/gamedev/zig-gamedev/libs/zmath/.zig-cache/o/9ce519e742c63ef799a10003aafec9b8/zmath-tests --listen=- 
```

Example (with this change):
```
test
└─ run zmath-tests 79/84 passed, 1 failed, 4 skipped
error: 'zmath.test.zmath.round' failed: expected 1.38861e6, found 1.38891e6
/Users/chris/zig/0.13.0-dev.351+64ef45eb0/files/lib/std/testing.zig:93:17: 0x10493cec3 in expectEqualInner__anon_1354 (zmath-tests)
                return error.TestExpectedEqual;
                ^
/Users/chris/gamedev/zig-gamedev/libs/zmath/src/zmath.zig:4508:9: 0x10493d143 in expectVecEqual__anon_1304 (zmath-tests)
        try std.testing.expectEqual(expected[i], actual[i]);
        ^
/Users/chris/gamedev/zig-gamedev/libs/zmath/src/zmath.zig:955:5: 0x10494587f in test.zmath.round (zmath-tests)
    try expectVecEqual(v, f32x4(-1_388_610.0, 1_388_910.0, 1_388_109.0, 2_388_610.0));
    ^
error: while executing test 'util.test.zmath.util.mat.right', the following test command failed:
/Users/chris/gamedev/zig-gamedev/libs/zmath/.zig-cache/o/a7b0b4f22e1898eb4eadbbd6f5506c76/zmath-tests --listen=- 
```

dmurph commented:
See the top line:
`error: 'zmath.test.zmath.round' failed: expected 1.38861e6, found 1.38891e6`

This now prints the exact mismatch.

hazeycode commented:
> An improvement here would be to save all errors for the given vec, and then return them all somehow so all mismatches are printed. Not sure how to do that, happy to make that change though if you have guidance.

I don't think it's worth the effort. But if someone figures out how to do it cleanly, it'd be nice.

dmurph commented:
I removed approxEqAbs, but I still needed it for negation checks, so I renamed it to vecApproxEqAbs....

But at that point I think we should just keep the original to not break other potential users. What do you think?

hazeycode commented:
I agree. If we are keeping it, not worth renaming.

dmurph commented:
Done

hazeycode commented:
Thanks!
